Mathlib/Algebra/Star/StarRingHom.lean

```diff
@@ -0,0 +1,469 @@
+  rfl
+
+@[simp]
+theorem comp_apply (f : B →⋆ₙ+* C) (g : A →⋆ₙ+* B) (a : A) : comp f g a = f (g a) :=
+  rfl
+
+@[simp]
+theorem comp_assoc (f : C →⋆ₙ+* D) (g : B →⋆ₙ+* C) (h : A →⋆ₙ+* B) :
+    (f.comp g).comp h = f.comp (g.comp h) :=
+  rfl
+
+@[simp]
+theorem id_comp (f : A →⋆ₙ+* B) : (NonUnitalStarRingHom.id _).comp f = f :=
+  ext fun _ => rfl
+
+@[simp]
+theorem comp_id (f : A →⋆ₙ+* B) : f.comp (NonUnitalStarRingHom.id _) = f :=
+  ext fun _ => rfl
+
+instance : Monoid (A →⋆ₙ+* A) where
+  mul := comp
+  mul_assoc := comp_assoc
+  one := NonUnitalStarRingHom.id A
+  one_mul := id_comp
+  mul_one := comp_id
+
+@[simp]
+theorem coe_one : ((1 : A →⋆ₙ+* A) : A → A) = id :=
+  rfl
+
+theorem one_apply (a : A) : (1 : A →⋆ₙ+* A) a = a :=
+  rfl
+
+end Basic
+
+section Zero
+
+-- the `zero` requires extra type class assumptions because we need `star_zero`
+variable {A B C D : Type*}
+variable [NonUnitalNonAssocSemiring A] [StarAddMonoid A]
+variable [NonUnitalNonAssocSemiring B] [StarAddMonoid B]
+
+instance : Zero (A →⋆ₙ+* B) :=
+  ⟨{ (0 : NonUnitalRingHom A B) with map_star' := by simp }⟩
+
+instance : Inhabited (A →⋆ₙ+* B) :=
+  ⟨0⟩
+
+instance : MonoidWithZero (A →⋆ₙ+* A) where
+  zero_mul := fun _ => ext fun _ => rfl
+  mul_zero := fun f => ext fun _ => map_zero f
+
+@[simp]
+theorem coe_zero : ((0 : A →⋆ₙ+* B) : A → B) = 0 :=
+  rfl
+
+theorem zero_apply (a : A) : (0 : A →⋆ₙ+* B) a = 0 :=
+  rfl
+
+end Zero
+
+
+end NonUnitalStarRingHom
+
+/-! ### Star ring equivalences -/
+
+/-- A *⋆-ring* equivalence is an equivalence preserving addition, multiplication, and the star
+operation, which allows for considering both unital and non-unital equivalences with a single
+structure. -/
+structure StarRingEquiv (A B : Type*) [Add A] [Add B] [Mul A] [Mul B] [Star A] [Star B]
+    extends A ≃+* B where
+  /-- By definition, a ⋆-ring equivalence preserves the `star` operation. -/
+  map_star' : ∀ a : A, toFun (star a) = star (toFun a)
+
+@[inherit_doc] notation:25 A " ≃⋆+* " B => StarRingEquiv A B
+
+/-- Reinterpret a star ring equivalence as a `RingEquiv` by forgetting the interaction with the star
+operation. -/
+add_decl_doc StarRingEquiv.toRingEquiv
+
+/-- `StarRingEquivClass F A B` asserts `F` is a type of bundled ⋆-ring equivalences between `A` and
+`B`.
+You should also extend this typeclass when you extend `StarRingEquiv`. -/
+class StarRingEquivClass (F : Type*) (A B : outParam Type*)
+  [Add A] [Mul A] [Star A] [Add B] [Mul B] [Star B] [EquivLike F A B]
+  extends RingEquivClass F A B : Prop
+  where
+  /-- By definition, a ⋆-ring equivalence preserves the `star` operation. -/
+  map_star : ∀ (f : F) (a : A), f (star a) = star (f a)
+
+namespace StarRingEquivClass
+
+-- See note [lower instance priority]
+instance (priority := 50) {F A B : Type*} {_ : Add A} {_ : Mul A} {_ : Star A} {_ : Add B}
+    {_ : Mul B} {_ : Star B} [EquivLike F A B] [hF : StarRingEquivClass F A B] :
```

Comment:
@eric-wieser do we still need these parameters to be implicit, or is this now a porting artifact? I'm inclined to think it's the latter.

Reply:
Link to the porting note in the porting PR by @ADedecker : https://github.com/leanprover-community/mathlib4/pull/2603/files#diff-8017bfdc9f21755fb88462d9ba5909d1590cfe91939577ce589ef704a831df57R700

Reply:
Eric is okay with https://github.com/leanprover-community/mathlib4/pull/14110#pullrequestreview-2140003696 so I've made the corresponding changes here.